flink-runtime/src/test/java/org/apache/flink/runtime/rest/compatibility/CompatibilityRoutines.java

```diff
@@ -0,0 +1,270 @@
+
+	private static void assertCompatible(final PathParameterContainer old, final PathParameterContainer cur) {
+		for (final PathParameterContainer.PathParameter oldParam : old.pathParameters) {
+			if (cur.pathParameters.stream().noneMatch(param -> param.key.equals(oldParam.key))) {
+				Assert.fail(String.format(
+					"Existing Path parameter %s was removed.",
+					oldParam.key));
+			}
+		}
+		// contrary to other routines path parameters must be completely identical between versions, so we have to
+		// check both directions
+		for (final PathParameterContainer.PathParameter curParam : cur.pathParameters) {
+			if (old.pathParameters.stream().noneMatch(param -> param.key.equals(curParam.key))) {
+				Assert.fail(String.format(
+					"New path parameter %s was added.",
+					curParam.key));
+			}
+		}
+	}
+
+	private static void assertCompatible(final QueryParameterContainer old, final QueryParameterContainer cur) {
+		for (final QueryParameterContainer.QueryParameter oldParam : old.queryParameters) {
+			final Optional<QueryParameterContainer.QueryParameter> matchingParameter = cur.queryParameters.stream()
+				.filter(param -> param.key.equals(oldParam.key))
+				.findAny();
+
+			if (matchingParameter.isPresent()) {
+				final QueryParameterContainer.QueryParameter newParam = matchingParameter.get();
+				if (!oldParam.mandatory && newParam.mandatory) {
+					Assert.fail(String.format(
+						"Previously optional query parameter %s is now mandatory.",
+						oldParam.key));
+				}
+			} else {
+				Assert.fail(String.format(
+					"Query parameter %s was removed.",
+					oldParam.key));
+			}
+		}
+	}
+
+	private static JsonNode extractSchema(final Class<?> messageClass) {
+		try {
+			final JsonSchema schema = schemaGen.generateSchema(messageClass);
+
+			return OBJECT_MAPPER.valueToTree(schema);
+		} catch (JsonProcessingException e) {
+			throw new RuntimeException("Failed to generate message schema for class " + messageClass.getCanonicalName() + '.', e);
+		}
+	}
+
+	private static void assertCompatible(final JsonNode old, final JsonNode cur) {
+		final Deque<Tuple2<JsonNode, JsonNode>> stack = new ArrayDeque<>(8);
+		stack.add(Tuple2.of(old, cur));
+
+		while (!stack.isEmpty()) {
+			final Tuple2<JsonNode, JsonNode> propertyPair = stack.pop();
+
+			final JsonNode oldProperty = propertyPair.f0;
+			final JsonNode curProperty = propertyPair.f1;
+
+			Assert.assertNotNull("Field " + oldProperty + " was removed.", curProperty);
+
+			final String oldType = oldProperty.get("type").asText();
+			final String curType = curProperty.get("type").asText();
+
+			// 'any' as a type only occurs for properties where we have custom serialization code
+			// removing this custom code (and thus improving the schema) should be possible
+			// hence we only check equality for other types
+			if (!oldType.equals("any")) {
+				Assert.assertEquals(
+					String.format("Type of field was changed from '%s' to '%s'.", oldType, curType),
```

Comment:
From the perspective of an HTTP client, the type shouldn't matter. Renaming a Java class does not break compatibility.

Reply:
when you switch from object/array to string it does matter to http clients.

Reply:
Ok sorry, I was confusing `type` with `id`